<h1><center>6. Cat and dog - Breed classification</center></h1>

<p style='text-align: center;'> 
Jishnu Jeevan <br>
Department of Computer Science <br>
M.Tech Computer and Information Science <br>
jishnujeevan@cusat.ac.in <br>
</p>

<h2><center> Assignemt Objective</center></h2>
<p style='text-align: justify;'>
The data set contains pictures of cats and dog. The cat and dog belong to different breeds. Our objective is to classify the cat and dog to there repective breed. There are a total of 37 breeds, each having 200 pictures for each breed.
</p>

<h2>The data set contains the following information: </h2>
<p style='text-align: justify;'>
The following annotations are available for every image in the dataset: 
1. Species and breed name <br>
2. A tight bounding box (ROI) around the head of the animal <br>
3. A pixel level foreground-background segmentation (Trimap) <br>

The cat breeds and the total number of pictures for each breeds are:<br>
1. Abyssinian: 198 <br>
2. Bengal: 200 <br>
3. Birman: 200 <br>
4. Bombay: 184 <br>
5. British: 200 <br>
6. Egyptian: 190 <br>
7. Maine Coon: 200 <br>
8. Persian: 200 <br>
9. Ragdoll: 200 <br>
10. Russian Blue: 200 <br>
11. Siamese: 199 <br>
12. Sphynx: 200 <br>

The dog breeds and the toal number of pictures for each breeds are: <br>
1. American Bulldog: 200 
2. American Pit Bull Terrier: 200 
3. Basset Hound: 200 
4. Beagle: 200 
5. Boxer: 199 
6. Chihuahua: 200 
7. English Cocker Spaniel: 196
8. English Setter: 200
9. German Shorthaired: 200
10. Great Pyrenees: 200
11. Havanese: 200
12. Japanese Chin: 200
13. Keeshond: 199
14. Leonberger: 200
15. Miniature Pinscher: 200
16. Newfoundland: 196
17. Pomeranian: 200
18. Pug: 200
19. Saint Bernard: 200
20. Samoyed: 200
21. Scottish Terrier: 199
22. Shiba Inu: 200
23. Staffordshire Bull Terrier: 189
24. Wheaten Terrier: 200
25. Yorkshire Terrier: 200

The total numer of cats and dogs are: <br>
1. Cats: 2371
2. Dog: 4978
3. Total: 7349
</p>

### 1. Import the required libraries

In [2]:
# For image procssing
import cv2

# For matrix operation
import numpy as np

# For acessing files from path
import os

# For random and shuffle operations
from random import shuffle

# To show a progress bar
from tqdm import tqdm

# For checking types
import types

### 2. Let us set the path of the image directory, the size of the images (as all images are not of the same size), learning rate and the model name.

In [4]:
DIR = './images'

# Every image we have we are going to reshape it to 50 x 50
IMG_SIZE = 50

# Learning rate is 0.001
LR = 0.001

# Make a model name so we can save it and load it later
MODEL_NAME = "dog vs cat.model"

### 3. Data pre processing

#### 3a. We are going to convert the labels into one hot encoding, maually

In [3]:
# The lable of the image is of the form Abyssinian_1.jpg
def label_img(img):
    # Split the word by '.' and take the first word
    word_label = img.split('.')[0]
    
    # Create a list containig all the breeds
    breed_list =[
                    "Abyssinian_", "Bengal_", "Birman_", "Bombay_", "British_Shorthair_", "Egyptian_Mau_",
                    "Maine_Coon_", "Persian_", "Ragdoll_", "Russian_Blue_", "Siamese_", "Sphynx_",
                    "american_bulldog_","american_pit_bull_terrier_","basset_hound_","beagle_","boxer_",
                    "chihuahua_","english_cocker_spaniel_","english_setter_","german_shorthaired_","great_pyrenees_",
                    "havanese_","japanese_chin_","keeshond_","leonberger_","miniature_pinscher_",
                    "newfoundland_","pomeranian_","pug_","saint_bernard_","samoyed_",
                    "scottish_terrier_","shiba_inu_","staffordshire_bull_terrier_","wheaten_terrier_","yorkshire_terrier_"
                ]
    
    # iterate through the list
    for index, breed in enumerate(breed_list):
        # Check if the breed is in lable
        if breed in word_label:   
        
            # create a list with 37 0's
            one_hot = [0 for i in range(37)]

            # Set the appropriate index to 1, for that breed
            one_hot[index] = 1

            # Retur the one_hot encoded variable 
            return one_hot

#### 3b. We are going to convert the image file into a numpy array

In [4]:
def create_dataset():
    # Create a list to store the image and its label
    data = []
    
    # Let us check how many images are valid images
    # Total images in directory
    total_img = 0
    
    # Total valid images in directory
    valid_img = 0
    
    # Take every image in the directory
    for img in tqdm(os.listdir(DIR)):
        
        # Increment the number of images in directory
        total_img = total_img + 1
        
        # Get the label of the image
        label = label_img(img)
        
        # Find the full path of the image
        path = DIR + "/" + img
        
        # Convert it to grey scale
        img = cv2.imread(path, cv2.IMREAD_GRAYSCALE)
        
        # Chek if the image is a valid imaage file
        if type(img) is np.ndarray:
            
            # Resize the image
            img = cv2.resize(img, (IMG_SIZE,IMG_SIZE))
        
            # Append the image array and label
            data.append([np.array(img), np.array(label)])
            
            # Increment the number of valid images
            valid_img = valid_img + 1 
    
    # Shuffle the data set
    shuffle(data)
    
    # Save it as an npy file
    np.save('Data.npy', data)
    
    print("Total images in directory : ", total_img)
    print("Valid images in directory : ", valid_img)
    
    # Return the data set
    return data

### 4. Loading the data set

#### 4a. If you don't have an .npy file then create it by running this code

In [5]:
dataset = create_dataset()

100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 7390/7390 [04:45<00:00, 25.87it/s]


Total images in directory :  7390
Valid images in directory :  7384


#### 4b. If you do have a .npy file, load it using this code

In [5]:
dataset = np.load('Data.npy')

### 5. I am going to use a neural network to identify the images. It will have 5 convolutional layer, a fully connected layer and an output layer. It will run a 100 epochs

In [6]:
# To surpress warning messages
import warnings
warnings.filterwarnings('ignore')

# Reset the graph instance
import tensorflow as tf
tf.reset_default_graph()

# We are going use tflearn to create the network
import tflearn

# For convolutional and max pooling layer
from tflearn.layers.conv import conv_2d, max_pool_2d

from tflearn.layers.normalization import local_response_normalization

# For input layer and fully connected layer
from tflearn.layers.core import input_data, fully_connected, dropout

# For regression
from tflearn.layers.estimator import regression

# Input node
convnet = input_data(shape=[None, IMG_SIZE, IMG_SIZE, 1], name='input')

# First convolutional layer
convnet = conv_2d(convnet, 32, 5, activation='relu')
convnet = max_pool_2d(convnet, 5)

# Second convolutional layer
convnet = conv_2d(convnet, 64, 5, activation='relu')
convnet = max_pool_2d(convnet, 5)

# Third convolutional layer
convnet = conv_2d(convnet, 128, 5, activation='relu')
convnet = max_pool_2d(convnet, 5)

# Fourth convolutional layer
convnet = conv_2d(convnet, 64, 5, activation='relu')
convnet = max_pool_2d(convnet, 5)

# Fifth convolutional layer
convnet = conv_2d(convnet, 32, 5, activation='relu')
convnet = max_pool_2d(convnet, 5)

# Fully connected layer
convnet = fully_connected(convnet, 1024, activation='relu')
convnet = dropout(convnet, 0.8)

# Output layer
convnet = fully_connected(convnet, 37, activation='softmax')
convnet = regression(convnet, optimizer='adam', learning_rate=LR, loss='categorical_crossentropy', name='targets')

model = tflearn.DNN(convnet, tensorboard_dir='log')

curses is not supported on this machine (please install/reinstall curses for an optimal experience)







Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor
Instructions for updating:
Use tf.initializers.variance_scaling instead with distribution=uniform to get equivalent behavior.

Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor
Instructions for updating:
Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.

Instructions for updating:
keep_dims is deprecated, use keepdims instead

Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where



### 6. Let us now train the model

### 6a. If the model already exist, load the model

In [7]:
# If the model alreaady exist
if os.path.exists('{}.meta'.format(MODEL_NAME)):
    model.load(MODEL_NAME)
    print('model loaded!')

Instructions for updating:
Use standard file APIs to check for files with this prefix.
INFO:tensorflow:Restoring parameters from C:\Users\LENOVO\Desktop\M.Tech\S1\2. MLA\Assaingment\Assignment 6\dog vs cat.model
model loaded!


#### 6b. Let us split the data set to train the model

In [16]:
# Take all but last 500 for training
train = dataset[:-500]

# Take last 500 for testing
test = dataset[-500:]

# This is the image we are going to sue for training
X = np.array([i[0] for i in train]).reshape(-1,IMG_SIZE,IMG_SIZE,1)

# This is the label we are going to use for training
Y = [i[1] for i in train]

# This is the image we are going to use for testing
test_x = np.array([i[0] for i in test]).reshape(-1,IMG_SIZE,IMG_SIZE,1)

# This is the label we are going to use for testing
test_y = [i[1] for i in test]

# Fit the model
model.fit(X,Y, n_epoch=100, validation_set=(test_x, test_y),snapshot_step=500, show_metric=True, run_id=MODEL_NAME)

Training Step: 10799  | total loss: 0.53980 | time: 17.509s
| Adam | epoch: 100 | loss: 0.53980 - acc: 0.8246 -- iter: 6848/6884
Training Step: 10800  | total loss: 0.54697 | time: 18.680s
| Adam | epoch: 100 | loss: 0.54697 - acc: 0.8219 | val_loss: 8.91028 - val_acc: 0.1480 -- iter: 6884/6884
--


#### 6c. If we are happy with the model save it

In [18]:
model.save(MODEL_NAME)

INFO:tensorflow:C:\Users\LENOVO\Desktop\M.Tech\S1\2. MLA\Assaingment\Assignment 6\dog vs cat.model is not in all_model_checkpoint_paths. Manually adding it.


### 7. As we can see we get an accuracy of 82%.